src/machine/machine-varlink.c

```diff
@@ -375,3 +378,188 @@ int vl_method_kill(sd_varlink *link, sd_json_variant *parameters, sd_varlink_met
+                        return sd_varlink_error_invalid_parameter_name(link, "user");
+                if (!isempty(p.path) && !path_is_absolute(p.path))
+                        return sd_varlink_error_invalid_parameter_name(link, "path");
+                // TODO(ikruglov): args validation?
+                if (!strv_isempty(p.env) && !strv_env_is_valid(p.env))
+                        return sd_varlink_error_invalid_parameter_name(link, "env");
+
+                user = isempty(p.user) ? "root" : p.user;
+                path = isempty(p.path) ? machine_default_shell_path() : p.path;
+                args = isempty(p.path) ? machine_default_shell_args(user) : strv_isempty(p.args) ? strv_new(path) : TAKE_PTR(p.args);
+                if (!args)
+                        return -ENOMEM;
+
+                command_line = strv_join(args, " ");
+                if (!command_line)
+                        return -ENOMEM;
+        }
+
+        r = lookup_machine_by_name_or_pidref(link, manager, p.name, &p.pidref, &machine);
+        if (r == -ESRCH)
+                return sd_varlink_error(link, "io.systemd.Machine.NoSuchMachine", NULL);
+        if (r != 0)
+                return r;
+
+        r = varlink_verify_polkit_async(
+                        link,
+                        manager->bus,
+                        machine_open_polkit_action(p.mode, machine->class),
+                        machine_open_polkit_details(p.mode, machine->name, user, path, command_line),
+                        &manager->polkit_registry);
+        if (r <= 0)
+                return r;
+
+        master = machine_openpt(machine, O_RDWR|O_NOCTTY|O_CLOEXEC, &pty_name);
+        if (ERRNO_IS_NEG_NOT_SUPPORTED(master))
+                return sd_varlink_error(link, "io.systemd.Machine.NotSupported", NULL);
+        if (master < 0)
+                return log_debug_errno(master, "Failed to open pseudo terminal: %m");
+
+        switch (p.mode) {
+                case MACHINE_OPEN_MODE_LOGIN:
+                        r = machine_start_getty(machine, pty_name, /* error = */ NULL);
+                        if (r == -ENOENT)
+                                return sd_varlink_error(link, "io.systemd.Machine.NoSDBus", NULL);
+                        if (ERRNO_IS_NEG_NOT_SUPPORTED(r))
+                                return sd_varlink_error(link, "io.systemd.Machine.NotSupported", NULL);
+                        if (r < 0)
+                                return log_debug_errno(r, "Failed to start getty for machine '%s': %m", machine->name);
+
+                        break;
+
+                case MACHINE_OPEN_MODE_SHELL: {
+                        _cleanup_close_ int slave = -EBADF;
+
+                        slave = machine_open_terminal(machine, pty_name, O_RDWR|O_NOCTTY|O_CLOEXEC);
+                        if (slave < 0)
+                                return log_debug_errno(slave, "Failed to open terminal: %m");
```

Comment:
maybe fow new code just avoid the "slave" term, to avoid fuss about it. just replace "slave" with "pty_fd" or so here, it's equally descriptive.

Reply:
and maybe renae "master" to "ptmx_fd"...